pkg/loop/internal/relayer/pluginprovider/chainreader/chain_reader_test.go

```diff
@@ -64,99 +64,148 @@ func TestVersionedBytesFunctions(t *testing.T) {
-	crTester := chainreadertest.WrapChainReaderTesterForLoop(&fakeChainReaderInterfaceTester{impl: impl})
-	crTester.Setup(t)
-	cr := crTester.GetChainReader(t)
+				ctx := tests.Context(t)
 
-	es := &errChainReader{}
-	errTester := chainreadertest.WrapChainReaderTesterForLoop(&fakeChainReaderInterfaceTester{impl: es})
-	errTester.Setup(t)
-	chainReader := errTester.GetChainReader(t)
+				nilTester := chainreadertest.WrapChainReaderTesterForLoop(&fakeChainReaderInterfaceTester{impl: nil})
+				nilTester.Setup(t)
+				nilCr := nilTester.GetChainReader(t)
 
-	t.Run("nil reader should return unimplemented", func(t *testing.T) {
-		ctx := tests.Context(t)
+				err := nilCr.GetLatestValue(ctx, "", "method", "anything", "anything")
+				assert.Equal(t, codes.Unimplemented, status.Convert(err).Code())
+			})
 
-		nilTester := chainreadertest.WrapChainReaderTesterForLoop(&fakeChainReaderInterfaceTester{impl: nil})
-		nilTester.Setup(t)
-		nilCr := nilTester.GetChainReader(t)
+			for _, errorType := range errorTypes {
+				es.err = errorType
+				t.Run("GetLatestValue unwraps errors from server "+errorType.Error(), func(t *testing.T) {
+					ctx := tests.Context(t)
+					err := chainReader.GetLatestValue(ctx, "", "method", nil, "anything")
+					assert.True(t, errors.Is(err, errorType))
+				})
+			}
 
-		_, err := nilCr.QueryKey(ctx, "", query.KeyFilter{}, query.LimitAndSort{}, &[]interface{}{nil})
-		assert.Equal(t, codes.Unimplemented, status.Convert(err).Code())
+			// make sure that errors come from client directly
+			es.err = nil
+			t.Run("GetLatestValue returns error if type cannot be encoded in the wire format", func(t *testing.T) {
+				ctx := tests.Context(t)
+				err := chainReader.GetLatestValue(ctx, "", "method", &cannotEncode{}, &TestStruct{})
+				assert.True(t, errors.Is(err, types.ErrInvalidType))
+			})
+		}
 	})
+}
 
-	for _, errorType := range errorTypes {
-		es.err = errorType
-		t.Run("QueryKey unwraps errors from server "+errorType.Error(), func(t *testing.T) {
-			ctx := tests.Context(t)
-			_, err := chainReader.QueryKey(ctx, "", query.KeyFilter{}, query.LimitAndSort{}, &[]interface{}{nil})
-			assert.True(t, errors.Is(err, errorType))
-		})
-	}
+func TestQueryKey(t *testing.T) {
+	t.Parallel()
+
+	chainreadertest.LoopEncodingTestMatrix(t, func(version chainreader.EncodingVersion) func(t *testing.T) {
+		return func(t *testing.T) {
+			t.Parallel()
+
+			impl := &protoConversionTestChainReader{}
+			crTester := chainreadertest.WrapChainReaderTesterForLoop(&fakeChainReaderInterfaceTester{impl: impl}, chainreadertest.WithChainReaderLoopEncoding(version))
+			crTester.Setup(t)
+			cr := crTester.GetChainReader(t)
+
+			es := &errChainReader{}
+			errTester := chainreadertest.WrapChainReaderTesterForLoop(&fakeChainReaderInterfaceTester{impl: es})
+			errTester.Setup(t)
+			chainReader := errTester.GetChainReader(t)
+
+			t.Run("nil reader should return unimplemented", func(t *testing.T) {
+				t.Parallel()
```

Comment:
This parallel call seems out of place since the other sub-tests under this parent test do not have parallel. IIUC this test would just defer until the non-parallel siblings are complete.